system/lib/wasmfs/syscalls.cpp

```diff
@@ -802,204 +767,149 @@ long __syscall_getdents64(long fd, long dirp, long count) {
-  // to lock the directories in reverse order and could cause deadlock.
-
-  // Edge case: rename("dir", "dir/somename") - in this scenario it should not
-  // be possible to rename the destination if the source is an ancestor.
-
-  // Edge case: rename("dir/somename", "dir") - in this scenario it should not
-  // be possible to rename the destination since by definition it is
-  // non-empty.
-
-  auto oldPathParts = splitPath((char*)old_path);
-
-  // For this operation to be atomic we must lock the source parent directory.
-  // getParsedPath() will return a locked parent directory.
-  long err;
-  auto oldParsedPath = getParsedPath(oldPathParts, err);
-
-  if (!oldParsedPath.parent) {
+  // Rename is the only syscall that needs to (or is allowed to) acquire locks
+  // on two directories at once. It requires locks on both the old and new
+  // parent directories to ensure that the moved file can be atomically removed
+  // from the old directory and added to the new directory without something
+  // changing that would prevent the move.
+  //
+  // To prevent deadlock in the case of simultaneous renames, serialize renames
+  // with an additional global lock.
+  static std::mutex renameMutex;
+  std::lock_guard<std::mutex> renameLock(renameMutex);
+
+  // Get the old directory.
+  auto parsedOld = path::parseParent((char*)old_path);
+  if (auto err = parsedOld.getError()) {
     return err;
   }
+  auto& [oldParent, oldFileNameView] = parsedOld.getParentChild();
+  std::string oldFileName(oldFileNameView);
 
-  if (oldPathParts.back().size() > WASMFS_NAME_MAX) {
-    return -ENAMETOOLONG;
-  }
-
-  if (!oldParsedPath.child) {
-    return -ENOENT;
-  }
-
-  // In Linux and WasmFS, renaming the root directory should return EBUSY.
-  // In the JS file system it reports EINVAL.
-  if (oldParsedPath.child == wasmFS.getRootDirectory()) {
-    return -EBUSY;
-  }
-
-  // Obtain the destination parent directory to see if it exists.
-  auto newPathParts = splitPath((char*)new_path);
-
-  if (newPathParts.empty()) {
-    return -ENOENT;
-  }
-
-  // In Linux, renaming a directory to the root directory returns ENOTEMPTY.
-  // TODO: Fix this when path parsing is refactored.
-  std::vector<std::string> root = {"/"};
-  if (newPathParts == root) {
-    return -ENOTEMPTY;
+  // Get the new directory.
+  auto parsedNew = path::parseParent((char*)new_path);
+  if (auto err = parsedNew.getError()) {
+    return err;
   }
+  auto& [newParent, newFileNameView] = parsedNew.getParentChild();
+  std::string newFileName(newFileNameView);
 
-  auto newBase = newPathParts.back();
-  if (newBase.size() > WASMFS_NAME_MAX) {
+  if (newFileNameView.size() > WASMFS_NAME_MAX) {
     return -ENAMETOOLONG;
   }
 
-  // oldParsedPath.child is the forbidden ancestor.
-  auto newParentDir = getDir(
-    newPathParts.begin(), newPathParts.end() - 1, err, oldParsedPath.child);
+  // Lock both directories. TODO: Avoid deadlocks here.
```

Comment:
doesn't the renameLock avoid this as the comment above says?

Reply:
Oh yes, stale comment.